src/infill/LightningLayer.cpp

```diff
@@ -0,0 +1,282 @@
+//Copyright (c) 2021 Ultimaker B.V.
+//CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#include "LightningLayer.h"
+
+#include <iterator> // advance
+
+#include "LightningTree.h"
+
+#include "../sliceDataStorage.h"
+#include "../utils/linearAlg2D.h"
+#include "../utils/SVG.h"
+#include "../utils/SparsePointGridInclusive.h"
+
+using namespace cura;
+
+coord_t LightningLayer::getWeightedDistance(const Point& boundary_loc, const Point& unsupported_loc)
+{
+    return vSize(boundary_loc - unsupported_loc);
+}
+
+LightningDistanceField::LightningDistanceField
+(
+    const coord_t& radius,
+ const Polygons& current_outline,
+ const Polygons& current_overhang,
+ const std::vector<std::shared_ptr<LightningTreeNode>>& initial_trees
+)
+: cell_size(radius / 6) // TODO: make configurable!
+, grid(cell_size)
+, supporting_radius(radius)
+, current_outline(current_outline)
+, current_overhang(current_overhang)
+{
+    std::vector<Point> regular_dots = PolygonUtils::spreadDotsArea(current_overhang, cell_size);
+    for (Point p : regular_dots)
+    {
+        const ClosestPolygonPoint cpp = PolygonUtils::findClosest(p, current_outline);
+        const coord_t dist_to_boundary = vSize(p - cpp.p());
+        unsupported_points.emplace_back(p, dist_to_boundary);
+    }
+    unsupported_points.sort(
+        [](const UnsupCell& a, const UnsupCell& b)
+        {
+            coord_t da = a.dist_to_boundary + std::hash<Point>{}(a.loc) % 191;
+            coord_t db = b.dist_to_boundary + std::hash<Point>{}(b.loc) % 191;
+            return da < db;
+        });
+    for (auto it = unsupported_points.begin(); it != unsupported_points.end(); ++it)
+    {
+        UnsupCell& cell = *it;
+        unsupported_points_grid.emplace(grid.toGridPoint(cell.loc), it);
+    }
+}
+
+bool LightningDistanceField::tryGetNextPoint(Point* p) const
+{
+    if (unsupported_points.empty()) return false;
+    *p = unsupported_points.front().loc;
+    return true;
+}
+
+void LightningDistanceField::update(const Point& to_node, const Point& added_leaf)
+{
+    auto process_func = 
+        [added_leaf, this](const SquareGrid::GridPoint& grid_loc)
+        {
+            auto it = unsupported_points_grid.find(grid_loc);
+            if (it != unsupported_points_grid.end())
+            {
+                std::list<UnsupCell>::iterator& list_it = it->second;
+                UnsupCell& cell = *list_it;
+                if (shorterThen(cell.loc - added_leaf, supporting_radius))
+                {
+                    unsupported_points.erase(list_it);
+                    unsupported_points_grid.erase(it);
+                }
+            }
+            return true;
+        };
+    const Point a = to_node;
+    const Point b = added_leaf;
+    Point ab = b - a;
+    Point ab_T = turn90CCW(ab);
+    Point extent = normal(ab_T, supporting_radius);
+    // TODO: process cells only once; make use of PolygonUtils::spreadDotsArea
```

Comment:
> @rburema remember what we discussed:
> 
> > all of my comments unless stated otherwise, can be ignored,

Sorry! TBF, there are a number I ignored _without_ doing the annoying thing and mailing back!

Reply:
That is okay @rburema I trust your decisions and the code is feedback enough